---
title: "Spatial quantile normalization for co-expression analysis"
shorttitle: "spqn guide"
author:
  - Yi Wang
  - Kasper D. Hansen
  - Stephanie C. Hicks
package: SpQN
abstract: >
  A guide to using spqn for co-expression analysis
vignette: >
  %\VignetteIndexEntry{spqn User's Guide}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
output:
  BiocStyle::html_document:
    toc_float: true
---


The estimation of pairwise correlations in the gene expression data is commonly used in the co-expression analysis such as network reconstruction of regulatory pathways. Our study reveals a dependency between the Pearson correlation and gene expression levels in bulk RNA-seq data, that highly expressed genes have highly variant correlations, which we refer to as mean-corelation relationship. Such dependency introduces bias in the downstream analysis that the co-expression signals among lowly expressed genes are likely to be ignored, such as transcription factors. To address this problem, we develop a method called spatial quantile normalization (SpQN), to normalize the the correlation matrix of gene expression data. We show that SpQN removes the mean-correlation relationship and corrects the expression bias in co-expression analysis.


# Preliminaries


Let's load the package and some example data.

```{r load, message=FALSE}
library(spqn)
library(spqnData)
library(SummarizedExperiment)
data(gtex.4k)
```

This dataset contains 4,000 genes from GTEx tissue "Adipose Subcutaneous" which we use as exemplar tissue in our preprint CITE. The data is from GTEx v6p. The genes have been restricted to a random sample of 4,000 genes from genes which are
- Either protein-coding or linRNAs
- Expressed (median expression greater than 0 on the $\log_2(\text{RPKM})$ scale

The data is counts (from the GTEx website) which has been transformed to $\log_2(\text{RPKM})$ and 4 principal components have been removed from the data matrix. Removing principal components from the data matrix also results in mean-centering. For this reason we need access to the original log-RPKM values (prior to removing principal components); these are stored in `rowData(gtex.4k)$ave_logrpkm`

First we sort the genes according to expression level and then calculate the correlation matrix
```{r cor_m, message=FALSE}
gtex.4k <- gtex.4k[order(rowData(gtex.4k)$ave_logrpkm),]
cor_m <- cor(t(assay(gtex.4k)))
```

# Check mean-correlation relationship in the data

We now recreate a number of diagnostic plots which illustrate the mean-correlation relationship

First compare the distribution of correlations in each expression level
```{r, message = F, fig.width = 4, fig.height = 4}
plot_density_condition_exp(cor_m)
```

Compare the distribution of correlations in signals and background, stratified by expression level
```{r, message = F, fig.width = 4, fig.height = 4}
plot_signal_condition_exp(cor_m, 0.1)
```

We now compute IQRs across the correlation matrix and plot them


Calculate IQR of correlations for each expression level, and compare the IQR of correlations between genes in each expression level
```{r, message=FALSE}
IQR_list <- get_IQR_condition_exp(cor_m, rowData(gtex.4k)$ave_logrpkm)
boxplot(IQR_list$IQR_cor_mat)
```


Compare the distribution of each bin and the reference bin, using qqplot
```{r, message = FALSE, fig.width = 3, fig.height = 3}
for(j in 1:10){
    qqplot_condition_exp(cor_m, j, j)
}
```

Find the relationship between the lowest of the two expression bins and IQR 
```{r, message = FALSE, fig.width = 4, fig.height = 4}
IQR_unlist <- unlist(lapply(1:10, function(ii) IQR_list$IQR_cor_mat[ii, ii:10]))
plot(rep(IQR_list$grp_mean, times = 1:10),
     IQR_unlist,
     xlab="min(average(log2CPM))", ylab="IQR", cex.lab=1.5, cex.axis=1.2, col="blue")
```

# To correct the mean-correlation dependency, apply SPQN in the correlation matrix, such that correlations of genes follows approximately the same distribution across different expression levels.
```{r spqn, message = FALSE, fig.width = 4, fig.height = 4}
cor_m_spqn <- normalize_correlation(cor_m, ngrp=20, size_grp=300, ref_grp=18)
```

# Check the mean-correlation dependency after applying SPQN.
Compare the distribution of correlations in each expression level
```{r, message = F, fig.width = 4, fig.height = 4}
plot_density_condition_exp(cor_m_spqn)
```

Compare the distribution of correlations in signals and background, stratified by expression level
```{r, message = F, fig.width = 4, fig.height = 4}
plot_signal_condition_exp(cor_m_spqn, 0.1)
```

Calculate IQR of correlations for each expression level, and compare the IQR of correlations between genes in each expression level
```{r, message=FALSE}
IQR_spqn_list <- get_IQR_condition_exp(cor_m_spqn, rowData(gtex.4k)$ave_logrpkm)
boxplot(IQR_spqn_list$IQR_cor_mat)
```


Compare the distribution of each bin and the reference bin, using qqplot
```{r, message = FALSE, fig.width = 3, fig.height = 3}
for(j in 1:10){
    qqplot_condition_exp(cor_m_spqn, j, j)
}
```

Find the relationship between the lowest of the two expression bins and IQR 
```{r, message = FALSE, fig.width = 4, fig.height = 4}
IQR_unlist <- unlist(lapply(1:10, function(ii) IQR_spqnlist$IQR_cor_mat[ii, ii:10]))
plot(rep(IQR_spqn_list$grp_mean, times = 1:10),
     IQR_unlist,
     xlab="min(average(log2CPM))", ylab="IQR", cex.lab=1.5, cex.axis=1.2, col="blue")
```


